include/SampleFrame.h

```diff
@@ -0,0 +1,234 @@
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#pragma once
+
+#ifndef SAMPLEFRAME_H
+#define SAMPLEFRAME_H
+
+#include "lmms_basics.h"
+
+#include <array>
+
+#include <cmath>
+#include <algorithm>
+
+
+namespace lmms
+{
+
+class SampleFrame
+{
+public:
+	SampleFrame() : SampleFrame(0., 0.)
+	{
+	}
+
+	SampleFrame(sample_t value) : SampleFrame(value, value)
+	{
+	}
+
+	SampleFrame(sample_t left, sample_t right) :
+		m_samples({ left, right })
+	{
+	}
+
+	sample_t* data()
+	{
+		return m_samples.data();
+	}
+
+	const sample_t* data() const
+	{
+		return m_samples.data();
+	}
+
+	sample_t& left()
+	{
+		return m_samples[0];
+	}
+
+	const sample_t& left() const
+	{
+		return m_samples[0];
+	}
+
+	void setLeft(const sample_t& value)
+	{
+		m_samples[0] = value;
+	}
+
+	sample_t& right()
+	{
+		return m_samples[1];
+	}
+
+	const sample_t& right() const
+	{
+		return m_samples[1];
+	}
+
+	void setRight(const sample_t& value)
+	{
+		m_samples[1] = value;
+	}
+
+	sample_t& operator[](size_t index)
+	{
+		return m_samples[index];
+	}
+
+	const sample_t& operator[](size_t index) const
+	{
+		return m_samples[index];
+	}
+
+	SampleFrame& operator=(float v)
+	{
+		left() = v;
+		right() = v;
+
+		return *this;
+	}
+
+	SampleFrame operator+(const SampleFrame& other) const
+	{
+		return SampleFrame(left() + other.left(), right() + other.right());
+	}
+
+	void operator+=(const SampleFrame& other)
+	{
+		auto & l = left();
+		auto & r = right();
+
+		l += other.left();
+		r += other.right();
+	}
+
+	SampleFrame operator*(float value) const
+	{
+		return SampleFrame(left() * value, right() * value);
+	}
+
+	void operator*=(float value)
+	{
+		setLeft(left() * value);
+		setRight(right() * value);
+	}
+
+	SampleFrame operator*(const SampleFrame& other) const
+	{
+		return SampleFrame(left() * other.left(), right() * other.right());
+	}
+
+	void operator*=(const SampleFrame& other)
+	{
+		left() *= other.left();
+		right() *= other.right();
+	}
+
+	sample_t scalarProduct(const SampleFrame& other) const
+	{
+		return left() * other.left() + right() * other.right();
+	}
+
+	SampleFrame abs() const
+	{
+		return SampleFrame{std::abs(this->left()), std::abs(this->right())};
+	}
+
+	void max(const SampleFrame& other)
+	{
+		if (other.left() > left())
+		{
+			setLeft(other.left());
+		}
+
+		if (other.right() > right())
+		{
+			setRight(other.right());
+		}
+	}
```

Comment:
This is another operation that seems too low-level and out of place. Samples are typically centred around zero, so finding the maximum in the positive direction is an unusual operation. This is in practice used in conjunction with `abs()` to find peak values, so perhaps an operation focused on that would make more sense. (Interestingly, `DelayEffect` computes its peaks without using `abs` - this seems like a bug to me, and not behaviour worth preserving.)

Reply:
Adjusted with commit 4b5ede54. `SampleFrame` now has an `absMax` method which is used by the delay.

A helper function to compute the absolute peak values of a buffer is introduced as well. This enables the removal of `AudioEngine::getPeakValues` which is not really the business of an audio engine anyway.